src/marlin_pc/constraints.rs

```diff
@@ -0,0 +1,1978 @@
+            degree_bounds_and_shift_powers: self.degree_bounds_and_shift_powers.clone(),
+        }
+    }
+}
+
+impl<CycleE, PG> AllocVar<VerifierKey<CycleE::E2>, <CycleE::E1 as PairingEngine>::Fr>
+    for VerifierKeyVar<CycleE, PG>
+where
+    CycleE: CycleEngine,
+    PG: PairingVar<CycleE::E2, <CycleE::E1 as PairingEngine>::Fr>,
+    <CycleE::E2 as PairingEngine>::G1Projective: MulAssign<<CycleE::E1 as PairingEngine>::Fq>,
+    <CycleE::E2 as PairingEngine>::G2Projective: MulAssign<<CycleE::E1 as PairingEngine>::Fq>,
+    <CycleE::E2 as PairingEngine>::G1Affine:
+        ToConstraintField<<<CycleE::E1 as PairingEngine>::Fr as Field>::BasePrimeField>,
+    <CycleE::E2 as PairingEngine>::G2Affine:
+        ToConstraintField<<<CycleE::E1 as PairingEngine>::Fr as Field>::BasePrimeField>,
+{
+    #[tracing::instrument(target = "r1cs", skip(cs, val))]
+    fn new_variable<T>(
+        cs: impl Into<Namespace<<CycleE::E1 as PairingEngine>::Fr>>,
+        val: impl FnOnce() -> Result<T, SynthesisError>,
+        mode: AllocationMode,
+    ) -> R1CSResult<Self>
+    where
+        T: Borrow<VerifierKey<CycleE::E2>>,
+    {
+        let vk_orig = val()?.borrow().clone();
+
+        let ns = cs.into();
+        let cs = ns.cs();
+
+        let VerifierKey {
+            vk,
+            degree_bounds_and_shift_powers,
+            ..
+        } = vk_orig;
+
+        let degree_bounds_and_shift_powers = degree_bounds_and_shift_powers.map(|vec| {
+            vec.iter()
+                .map(|(s, g)| {
+                    (
+                        *s,
+                        FpVar::<<CycleE::E1 as PairingEngine>::Fr>::new_variable(
+                            ark_relations::ns!(cs, "degree bound"),
+                            || {
+                                Ok(<<CycleE::E1 as PairingEngine>::Fr as From<u128>>::from(
+                                    *s as u128,
+                                ))
+                            },
+                            mode,
+                        )
+                        .unwrap(),
+                        PG::G1Var::new_variable(ark_relations::ns!(cs, "pow"), || Ok(*g), mode)
+                            .unwrap(),
+                    )
+                })
+                .collect()
+        });
+
+        let KZG10VerifierKey { g, h, beta_h, .. } = vk;
+
+        let g = PG::G1Var::new_variable(ark_relations::ns!(cs, "g"), || Ok(g), mode)?;
+        let h = PG::G2Var::new_variable(ark_relations::ns!(cs, "h"), || Ok(h), mode)?;
+        let beta_h =
+            PG::G2Var::new_variable(ark_relations::ns!(cs, "beta_h"), || Ok(beta_h), mode)?;
+
+        Ok(Self {
+            g,
+            h,
+            beta_h,
+            degree_bounds_and_shift_powers,
+        })
+    }
+}
+
+impl<CycleE, PG> ToBytesGadget<<CycleE::E1 as PairingEngine>::Fr> for VerifierKeyVar<CycleE, PG>
+where
+    CycleE: CycleEngine,
+
+    PG: PairingVar<CycleE::E2, <CycleE::E1 as PairingEngine>::Fr>,
```

Comment:
I would add a type as follows:

`type E1Fr<CycleE> = <CycleE::E1 as PairingEngine>::Fr;`

And similarly for the other types that you use. It could help clean up some of the logic in the bounds. If you want to shorten it further, I would recommend changing `CycleE` to `E` in all the generic parameters everywhere.